### PR TITLE
[AC-2172] Member modal - limit admin access

### PR DESCRIPTION
**Pull request by eliykat**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

In the Member modal, if owners and admins cannot manage all collections and items, then you should only be able to edit collection assignments for collections that you have Can Manage access to.

This depends on:
* organization getter logic being updated in #8394 (stacked on that PR)
* server response model changes in bitwarden/server#3793

On the plus side, once those are merged, this PR is fairly minimal (if dense).

Sever changes are WIP.

## Code changes

* `member-dialog.component.ts` - most of this diff is just adding a `collection.canEdit` check in `mapCollectionToAccessItemView`. However, that requires the org and the v1 flag, so there's a large footprint just getting those values there.
  * note that this (should) provide us with optionality for **both** the v1 flag and the collection management setting, because everything hinges off this readonly property.
* `access-selector.component.ts` - 
  * there were a few places where we used `item.readonly` to mean "collection access via a group", because that was the only type of readonly item. However, we now have 2 types (collection access via group, and a collection you don't manage), so some places needed updating to limit their reach.
  * when giving the multiselect its initial items (via the selectionList), exclude any readonly items, because you can't give access to something that's readonly.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts [hunk 1]
```diff
@@ -581,7 +595,7 @@ function mapCollectionToAccessItemView(
     id: group ? `${collection.id}-${group.id}` : collection.id,
     labelName: collection.name,
     listName: collection.name,
-    readonly: group !== undefined,
+    readonly: group !== undefined || !collection.manage,
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts:598]: TODO: needs to use the `collection.canEdit` method to take into account collection access settings etc. That method itself may need to be updated.

File: apps/web/src/app/admin-console/organizations/shared/components/access-selector/access-selector.component.ts [hunk 2]
```diff
@@ -328,6 +330,14 @@ export class AccessSelectorComponent implements ControlValueAccessor, OnInit, On
     return this.permissionMode == PermissionMode.Edit && !item.readonly && !item.accessAllItems;
   }
 
+  /**
+   * Deselected items that are capable of being edited/selected. Excludes readonly items that the user does not have
+   * permission to grant access to.
+   */
+  protected get readonlyDeselectedItems() {
```
Comment (vincentsalucci) [inline apps/web/src/app/admin-console/organizations/shared/components/access-selector/access-selector.component.ts:337]: I think this name is confusing me. The description says this is for items that **are** capable of being edited/selected, but the name says they're `readonly` deselected items. Am I just misunderstanding?

Comment (shane-melton) [reply apps/web/src/app/admin-console/organizations/shared/components/access-selector/access-selector.component.ts:337]: Agreed, I think this should be something like: `editableDeselectedItems`

File: apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts [hunk 3]
```diff
@@ -206,25 +207,44 @@ export class MemberDialogComponent implements OnDestroy {
...
+          if (this.params.organizationUserId) {
+            this.loadOrganizationUser(
+              userDetails,
+              groups,
+              collections,
+              organization,
+              flexibleCollectionsV1Enabled,
+            );
+          } else {
+            // Populate all available collections, without any user details
```
Comment (shane-melton) [inline apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts:239]: question: I'm probably just missing something obvious, but why do we have this `else {}` clause now?

**Review by shane-melton — COMMENTED** [review]
This looks pretty good! Just a question and a comment on the new `readonlyDeselectedItems` property name.

File: apps/web/src/app/admin-console/organizations/shared/components/access-selector/access-selector.component.ts (hunk 2 above)
Comment (eliykat) [reply apps/web/src/app/admin-console/organizations/shared/components/access-selector/access-selector.component.ts:337]: Fixed!

File: apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts (hunk 3 above)
Comment (eliykat) [reply apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts:239]: You're right that it's duplicative, but I think it's the earlier assignment that needs to be removed. `loadOrganizationUser` [sets this.collectionAccessItems from scratch](https://github.com/bitwarden/clients/blob/ac/ac-2172/member-modal---access-selector-changes/apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts#L320-L328) which will overwrite whatever's set earlier in this subscribe callback. So I'll remove the earlier setting of this property and we can rely on the if/else to set it depending on whether there's an orgUser or not.

**Comment by eliykat** [comment]
I had some interesting issues with data flow in Angular. I noticed that disabled rows were still appearing in the value being emitted by the access-selector until the user interacted with it. (The commit exhibiting this problem is https://github.com/bitwarden/clients/pull/8343/commits/3d406bb2d07bec091fddc8271d55649cf0ea882b.)

I tried to fix this in https://github.com/bitwarden/clients/pull/8343/commits/fdbeb4863807a437390171495f6862f78271da10, but that is not a proper fix, and it caused unit tests to start failing because the control was emitting more times than expected.

My diagnosis:
* `writeValue` takes the FormControl value and updates the native HTML elements (unidirectional)
* the callback registered by `registerOnChange` takes the HTML values and updates the FormControl value (unidirectional)
* however, my implementation was giving potentially disabled values to `writeValue` (i.e. access selections that the user doesn't have permission to edit), and `writeValue` was then updating the HTML elements, but also expecting it to _write back to the FormControl value_ to remove these disabled items (bidirectional! bad)

After some experimentation, I thought the best way to resolve it was to not pass potentially "disabled" values into the FormControl to begin with. The fixes are in [the last 2 commits](https://github.com/bitwarden/clients/pull/8343/files/0dad7c2f505d1d3dcc66b50e21d091209a3eef90..9d4ce3b9053476d8533ef74be46560a24ca590d3). The updated logic is:
* the access-selector FormControl only emits enabled (non-readonly) rows, so you should only give it non-readonly rows (AccessItemValues). This avoids the bidirectional flow issue, because `writeValue` is given a valid value from the start and does not need to alter it, only set up the rendering
* the AccessItemViews you pass in should represent _selectable_ (non-readonly) or _already selected_ items
* all readonly AccessItemViews are selected immediately (this actually reverts previous changes so that all readonly views are treated the same, rather than treating groups differently)
* AccessItemViews that are readonly and not already assigned can never be selected, so they're never passed in

It does make the logic more complex, but the LOC impact is still small, and it avoids trying to hack around Angular change detection.

**Review by vincentsalucci — APPROVED** [review]

**Comment by eliykat** [comment]
I noticed that the previous logic did not filter out readonly collections when inviting a new user. This meant they were all selected by default. I moved the initial assignment of `collectionAccessItems` out of `loadOrganizationUser` and back into the main loading method so that it runs for both new and existing orgUsers, with some tweaks to allow for a `null` userDetails.

File: apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts [hunk 4]
```diff
@@ -295,13 +325,22 @@ export class MemberDialogComponent implements OnDestroy {
...
+          flexibleCollectionsV1Enabled,
+          accessSelection,
+          group,
+        ),
       ),
     );
 
-    const accessSelections = mapToAccessSelections(userDetails);
+    // Set current collections and groups the user has access to (excluding collections the current user doesn't have
+    // permissions to change - they are included as readonly via the CollectionAccessItems
```
Comment (vincentsalucci) [inline apps/web/src/app/admin-console/organizations/members/components/member-dialog/member-dialog.component.ts:342]: Missing a closing `)` for this comment.

**Review by vincentsalucci — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]

**Pull request merged**